File: y2017/rust/src/bin/day4.rs
use std::fs;
use std::collections::{HashSet, HashMap};

fn valid_phrases1<S: AsRef<str>>(phrase: S) -> bool {
    let mut seen = HashSet::new();
    for p in phrase.as_ref().split_whitespace() {
        if seen.contains(p) {
            return false;
        }
        seen.insert(p);
    }
    true
}

fn valid_phrases2<S: AsRef<str>>(phrase: S) -> bool {
    let phrase = phrase.as_ref().to_lowercase();
    assert!(phrase.chars().all(|c| c.is_alphabetic() || c.is_whitespace()));

    fn str_occ_num(s: &str) -> [u32; 26] {
        let mut occs = [0; 26];
        for c in s.chars() {
            occs[c as usize - 'a' as usize] += 1;
        }
        occs
    }

    let mut seen = HashSet::new();
    for p in phrase.split_whitespace() {
        if !seen.insert(str_occ_num(p)) {
            return false;
        }
    }
    true
}

fn main() {
    let lines = fs::read_to_string("../input/day4.txt").expect("bruh");

    let mut valid1 = 0;
    let mut valid2 = 0;
    for p in lines.lines() {
        valid1 += if valid_phrases1(p) { 1 } else { 0 };
        valid2 += if valid_phrases2(p) { 1 } else { 0 };
    }

    println!("# of valid passphrases (p1): {valid1}");
    println!("# of valid passphrases (p2): {valid2}");
}

File: y2017/nim/day4.nim
import strutils
import sequtils
import algorithm
import sets

proc validPhrases1(phrase: string): bool = 
  var found: HashSet[string]
  for p in phrase.splitWhitespace():
    if p in found:
      return false
    found.incl(p)
  return true

proc validPhrases2(phrase: string): bool =
  var found: HashSet[string]
  for p in phrase.splitWhitespace():
    var pChars = toSeq(p.items)
    sort(pChars, system.cmp)

    let sortedP = $pChars
    if sortedP in found:
      return false
    found.incl(sortedP)
  return true

var valid1 = 0
var valid2 = 0
readFile("input/day4.txt")
  .strip()
  .splitLines()
  .apply(proc(p: string) =
    valid1 += (if validPhrases1(ref p): 1 else: 0)
    valid2 += (if validPhrases2(ref p): 1 else: 0))

echo "# of valid passphrases (p1): ", valid1
echo "# of valid passphrases (p2): ", valid2

